### PR TITLE
Fix Clear Amulet visual bug + Tests

### LINKED ISSUE #3054: Triple Kick calls attackstring and ppreduce twice
### Description

The battle script for Triple Kick's effect calls ppreduce and attackstring.
Then it jumps to `BattleScript_HitFromCritCalc` which also calls both of them.

### Version

upcoming (Edge)

### Upcoming Version

8efbe825dbd7a306af3ee8c8b4cdd4608375573a

### Discord contact info

Lunos#4026

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Description
Fixes Clear Amulet visual bug. A wrong message was displayed when Intimidate on Clear Amulet was used.
Also removed some code in the Intimidate script that was not needed and added a couple of tests.

pokeemerald-5
pokeemerald-6

##  **Issue(s) that this PR fixes **
Fixes #3054

## **Discord contact info**
rainonline

### CODE REVIEW

File: test/ability_clear_body.c [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        HP_BAR(player, captureDamage: &turnOneHit);
+        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        ABILITY_POPUP(opponent, ABILITY_CLEAR_BODY);
+        MESSAGE("Foe Beldum's Clear Body prevents stat loss!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/ability_clear_body.c:27]:
```suggestion
}

TO_DO_BATTLE_TEST("Clear Body prevents stat stage reduction from moves"); // Growl, Leer, Confide, Fake Tears, Scary Face, Sweet Scent, Sand Attack (Attack, Defense, Sp. Attack, Sp. Defense, Speed, Evasion, Accuracy
TO_DO_BATTLE_TEST("Clear Body prevents Sticky Web");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent stat stage reduction from moves used by the user"); // e.g. Superpower
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Speed reduction from Iron Ball");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Speed reduction from paralysis");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Attack reduction from burn");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent receiving negative stat changes from Baton Pass");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Topsy-Turvy");
TO_DO_BATTLE_TEST("Clear Body doesn't prevent Spectral Thief from resetting positive stat changes");
TO_DO_BATTLE_TEST("Clear Body is ignored by Mold Breaker");
```

File: test/ability_white_smoke.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        HP_BAR(player, captureDamage: &turnOneHit);
+        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        ABILITY_POPUP(opponent, ABILITY_WHITE_SMOKE);
+        MESSAGE("Foe Torkoal's White Smoke prevents stat loss!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/ability_white_smoke.c:27]:
```suggestion
}

TO_DO_BATTLE_TEST("White Smoke prevents stat stage reduction from moves"); // Growl, Leer, Confide, Fake Tears, Scary Face, Sweet Scent, Sand Attack (Attack, Defense, Sp. Attack, Sp. Defense, Speed, Evasion, Accuracy
TO_DO_BATTLE_TEST("White Smoke prevents Sticky Web");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent stat stage reduction from moves used by the user"); // e.g. Superpower
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Speed reduction from Iron Ball");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Speed reduction from paralysis");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Attack reduction from burn");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent receiving negative stat changes from Baton Pass");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Topsy-Turvy");
TO_DO_BATTLE_TEST("White Smoke doesn't prevent Spectral Thief from resetting positive stat changes");
TO_DO_BATTLE_TEST("White Smoke is ignored by Mold Breaker");
```

File: test/ability_full_metal_body.c [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        HP_BAR(player, captureDamage: &turnOneHit);
+        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        ABILITY_POPUP(opponent, ABILITY_FULL_METAL_BODY);
+        MESSAGE("Foe Solgaleo's Full Metal Body prevents stat loss!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/ability_full_metal_body.c:27]:
```suggestion
}

TO_DO_BATTLE_TEST("Full Metal Body prevents stat stage reduction from moves"); // Growl, Leer, Confide, Fake Tears, Scary Face, Sweet Scent, Sand Attack (Attack, Defense, Sp. Attack, Sp. Defense, Speed, Evasion, Accuracy
TO_DO_BATTLE_TEST("Full Metal Body prevents Sticky Web");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent stat stage reduction from moves used by the user"); // e.g. Superpower
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Speed reduction from Iron Ball");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Speed reduction from paralysis");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Attack reduction from burn");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent receiving negative stat changes from Baton Pass");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Topsy-Turvy");
TO_DO_BATTLE_TEST("Full Metal Body doesn't prevent Spectral Thief from resetting positive stat changes");
TO_DO_BATTLE_TEST("Full Metal Body is ignored by Mold Breaker");
```

File: test/ability_inner_focus.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        HP_BAR(player, captureDamage: &turnOneHit);
+        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        ABILITY_POPUP(opponent, ABILITY_INNER_FOCUS);
+        MESSAGE("Foe Zubat's Attack was not lowered!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/ability_inner_focus.c:28]:
```suggestion
}

TO_DO_BATTLE_TEST("Inner Focus prevents flinching");
TO_DO_BATTLE_TEST("Inner Focus is ignored by Mold Breaker");
```

File: test/ability_own_tempo.c [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        HP_BAR(player, captureDamage: &turnOneHit);
+        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        ABILITY_POPUP(opponent, ABILITY_OWN_TEMPO);
+        MESSAGE("Foe Slowpoke's Attack was not lowered!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/ability_own_tempo.c:28]:
```suggestion
}

TO_DO_BATTLE_TEST("Own Tempo prevents confusion from moves by the opponent");
TO_DO_BATTLE_TEST("Own Tempo prevents confusion from moves by the user");
TO_DO_BATTLE_TEST("Own Tempo prevents confusion from items"); // e.g. Iapapa Berry
TO_DO_BATTLE_TEST("Own Tempo cures confusion if it's obtained via Skill Swap");
TO_DO_BATTLE_TEST("Own Tempo cures confusion obtained from an opponent with Mold Breaker");
```

File: test/ability_scrappy.c [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        HP_BAR(player, captureDamage: &turnOneHit);
+        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        ABILITY_POPUP(opponent, ABILITY_SCRAPPY);
+        MESSAGE("Foe Kangaskhan's Attack was not lowered!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/ability_scrappy.c:28]:
```suggestion
}

TO_DO_BATTLE_TEST("Scrappy allows to hit Ghost-type Pokémon with Normal- and Fighting-type moves");
TO_DO_BATTLE_TEST("Scrappy doesn't bypass a Ghost-type's Wonder Guard");
```

File: test/hold_effect_clear_amulet.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        MESSAGE("Foe Wobbuffet's Attack was not lowered!");
+        HP_BAR(player, captureDamage: &turnTwoHit);
+    } THEN {
+        EXPECT_EQ(turnOneHit, turnTwoHit);
+    }
+}
+
+SINGLE_BATTLE_TEST("Clear Amulet prevents stat reducing effects")
+{
```
Comment (AsparagusEduardo) [inline test/hold_effect_clear_amulet.c:33]: To make sure every stat change is handled, let's use `PARAMETRIZE` to add:
- Leer (Defense)
- Confide (Sp. Attack)
- Fake Tears (Sp. Defense)
- Scary Face (Speed)
- Sweet Scent (Evasion)
- Sand Attack (Accuracy)

File: test/hold_effect_clear_amulet.c [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_GROWL); }
+    } SCENE {
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, opponent); }
+        MESSAGE("Foe Wobbuffet's Attack was not lowered!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Clear Amulet prevents secondary effects that reduce stats")
+{
```
Comment (AsparagusEduardo) [inline test/hold_effect_clear_amulet.c:47]: To make sure every stat change is handled, let's use `PARAMETRIZE` to add:
- Aurora Beam (Attack)
- Snarl (Sp. Attack)
- Psychic (Sp. Defense)
- Bubble Beam (Speed)
- Mud Slap (Accuracy)

Unfortunately, there's no move for evasion that does this.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Comment by AlexOn1ine** [comment]
Added all the TODO tests, wrote some additional tests for Own Tempo, Inner Focus, Scrappy and added the missing Clear Amulet checks. Also fixed a tiny Triple Kick bug (issue #3054)

**Comment by PacFire** [comment]
swag

**Comment by AsparagusEduardo** [comment]
Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.

**Comment by AlexOn1ine** [comment]
> Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.

Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.

**Comment by mrgriffin** [comment]
> > Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.
> 
> Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.

Something I often do is write a message saying that I intend to rebase and force push to clean up the history once the PR is accepted :)

**Comment by AlexOn1ine** [comment]
> > > Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.
> > 
> > 
> > Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.
> 
> Something I often do as a PR author is write a message saying that I intend to rebase and force push to clean up the history once the PR is accepted :)

Good suggestion! 👍

**Comment by AsparagusEduardo** [comment]
> > > > Please avoid doing force pushes, it makes it harder to tell what was changed in a PR.
> > > 
> > > 
> > > Oh will avoid doing it, sorry. I have done that to have a clean commit history but I understand.
> > 
> > 
> > Something I often do as a PR author is write a message saying that I intend to rebase and force push to clean up the history once the PR is accepted :)
> 
> Good suggestion! 👍

Even then, if it's intended to be a single commit, we can squash it ourselves upon merging thanks to GitHub's Squash and Merge option :)

File: test/hold_effect_clear_amulet.c [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        else if (move == MOVE_CONFIDE)
+            MESSAGE("Foe Wobbuffet's Sp. Atk was not lowered!");
+        else if (move == MOVE_FAKE_TEARS)
+            MESSAGE("Foe Wobbuffet's Sp. Def was not lowered!");
+        else if (move == MOVE_SCARY_FACE)
+            MESSAGE("Foe Wobbuffet's Speed was not lowered!");
+        else if (move == MOVE_SWEET_SCENT)
+            MESSAGE("Foe Wobbuffet's evasiveness was not lowered!");
+        else if (move == MOVE_SAND_ATTACK)
+            MESSAGE("Foe Wobbuffet's accuracy was not lowered!");
```
Comment (AsparagusEduardo) [inline test/hold_effect_clear_amulet.c:72]: My instinct would be to convert this to a switch, but that would increase the vertical space. Hmm...

File: test/ability_pastel_veil.c [hunk 10]
```diff
@@ -33,18 +33,22 @@ DOUBLE_BATTLE_TEST("Pastel Veil prevents Poison Sting poison on partner")
...
+    KNOWN_FAILING;
     GIVEN {
         ASSUME(gBattleMoves[MOVE_TOXIC].effect == EFFECT_TOXIC);
         PLAYER(SPECIES_PINSIR) { Ability(ABILITY_MOLD_BREAKER); }
         OPPONENT(SPECIES_PONYTA_GALARIAN) { Ability(ABILITY_PASTEL_VEIL); }
     } WHEN {
-        TURN { MOVE(player, MOVE_TOXIC); }
+        TURN { MOVE(player, MOVE_TOXIC);
+               MOVE(opponent, MOVE_TACKLE);
+        }
```
Comment (AsparagusEduardo) [inline test/ability_pastel_veil.c:44]:
```suggestion
        TURN { MOVE(player, MOVE_TOXIC); MOVE(opponent, MOVE_TACKLE); }
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: test/hold_effect_clear_amulet.c (hunk 9 above)
Comment (AlexOn1ine) [reply test/hold_effect_clear_amulet.c:72]: True, switches here might make more sense.

**Comment by AlexOn1ine** [comment]
Added the Wonder Guard message, changes if to switch and fixed misalignment.

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**